internal/factsengine/gatherers/saplocalhostresolver.go

```diff
@@ -0,0 +1,177 @@
+		Message: "error reading the sap profiles file system",
+	}
+)
+
+//go:generate mockery --name=HostnameResolver
+type HostnameResolver interface {
+	LookupHost(host string) ([]string, error)
+}
+
+//go:generate mockery --name=HostPinger
+type HostPinger interface {
+	Ping(host string) bool
+}
+
+type SapLocalhostResolverGatherer struct {
+	fs afero.Fs
+	hr HostnameResolver
+	hp HostPinger
+}
+
+type ResolvabilityDetails struct {
+	Hostname     string   `json:"hostname"`
+	Addresses    []string `json:"addresses"`
+	InstanceName string   `json:"instance_name"`
+	Reachability bool     `json:"reachability"`
+}
+
+type Resolver struct{}
+
+type Pinger struct{}
+
+func (r *Resolver) LookupHost(host string) ([]string, error) {
+	return net.LookupHost(host)
+}
+
+func (p *Pinger) Ping(host string) bool {
+	pinger, err := probing.NewPinger(host)
+	if err != nil {
+		return false
+	}
+	pinger.Count = 3
+	err = pinger.Run()
+	return err == nil
+}
+
+func NewDefaultSapLocalhostResolverGatherer() *SapLocalhostResolverGatherer {
+	return NewSapLocalhostResolver(afero.NewOsFs(), &Resolver{}, &Pinger{})
+}
+
+func NewSapLocalhostResolver(fs afero.Fs, hr HostnameResolver, hp HostPinger) *SapLocalhostResolverGatherer {
+	return &SapLocalhostResolverGatherer{fs: fs, hr: hr, hp: hp}
+}
+
+func (r *SapLocalhostResolverGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := make([]entities.Fact, 0)
+
+	details, err := r.getInstanceHostnameDetails()
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+
+	for _, factReq := range factsRequests {
+		var fact entities.Fact
+		factValue, err := mapReachabilityDetailsToFactValue(details)
+		if err != nil {
+			log.Error(err)
+			fact = entities.NewFactGatheredWithError(factReq, SapLocalhostResolverGathererDecodingError.Wrap(err.Error()))
+		} else {
+			fact = entities.NewFactGatheredWithRequest(factReq, factValue)
+		}
+		facts = append(facts, fact)
+	}
+
+	return facts, nil
+}
+
+func (r *SapLocalhostResolverGatherer) getInstanceHostnameDetails() (map[string][]ResolvabilityDetails, error) {
+	systems, err := sapsystem.FindSystems(r.fs)
+	if err != nil {
+		return nil, err
+	}
+
+	resolvabilityDetails := make(map[string][]ResolvabilityDetails)
+	for _, system := range systems {
+		sid := filepath.Base(system)
+		profileFiles, err := sapsystem.FindProfiles(r.fs, sid)
+		if err != nil {
+			return nil, SapLocalhostResolverFileSystemError.Wrap(err.Error())
+		}
+
+		for _, profileFile := range profileFiles {
+			if profileFile == sapsystem.SapDefaultProfile {
+				continue
+			}
+
+			match := hostnameRegexCompiled.FindStringSubmatch(profileFile)
+			if len(match) != regexSubgroupsCount {
+				continue
+			}
+			matchedSID := match[1]
+			matchedInstanceName := match[2]
+			matchedHostname := match[3]
+
+			addresses, err := r.hr.LookupHost(matchedHostname)
+			if err != nil {
+				return nil, SapLocalhostResolverHostnameResolutionError.Wrap(err.Error())
```

Comment:
We shouldn't error in this case. We should set an empty list of addresses.
Otherwise, a failure in 1 hostname condemns the fact gathering to fail, while some other honstnames might be resolvable